From f80c580b98f10150e529c4b266b353a7b585d184 Mon Sep 17 00:00:00 2001
From: Danny Baumann <dannybaumann@web.de>
Date: Fri, 31 Aug 2018 13:24:14 +0200
Subject: [PATCH 60/67] Stop initializing app ops in Camera default
 constructor.

The app ops are only initialized for determining whether shutter sound
is to be played or not, but there's no shutter sound to be played
without an actual camera instance or not. Additionally, if the app op is
determined as 'denied', trying to disable the shutter sound will
inevitably crash, e.g. with the following trace:

Caused by: java.lang.RuntimeException: Camera is being used after Camera.release() was called
       at android.hardware.Camera._enableShutterSound(Native Method)
       at android.hardware.Camera.updateAppOpsPlayAudio(Camera.java:1770)
       at android.hardware.Camera.initAppOps(Camera.java:582)
       at android.hardware.Camera.<init>(Camera.java:575)
       at android.hardware.Camera.getEmptyParameters(Camera.java:2130)
       at android.hardware.camera2.legacy.LegacyMetadataMapper.createCharacteristics(LegacyMetadataMapper.java:151)
       at android.hardware.camera2.CameraManager.getCameraCharacteristics(CameraManager.java:274)
       <...>

This happens because in the default constructor case no camera instance
is ever initialized, so we shouldn't try to manipulate said instance.

Change-Id: If3c3b01a290c18da0983ec2cac8ac9037adf3ac6
---
 core/java/android/hardware/Camera.java | 1 -
 1 file changed, 1 deletion(-)

diff --git a/core/java/android/hardware/Camera.java b/core/java/android/hardware/Camera.java
index 9350aab799a..b504327c403 100644
--- a/core/java/android/hardware/Camera.java
+++ b/core/java/android/hardware/Camera.java
@@ -572,7 +572,6 @@ public class Camera {
      * An empty Camera for testing purpose.
      */
     Camera() {
-        initAppOps();
     }
 
     private void initAppOps() {
-- 
2.17.1

